src/nspawn/nspawn-mount.c

```diff
@@ -703,17 +707,18 @@ static int parse_mount_bind_options(const char *options, unsigned long *mount_fl
         return 0;
 }
 
-static int mount_bind(const char *dest, CustomMount *m) {
+static int mount_bind(const char *dest, CustomMount *m, uid_t uid_shift, uid_t uid_range) {
```

Comment:
I think there should be a version of `mount_bind()` or an additional argument to `mount_bind()` that lets the caller pass in a file descriptor for an already existing user namespace. Especially for the `systemd-nspawn` case where the rootfs of the container itself is idmapped this makes sense. As long as the custom mount is not supposed to run with an idmapping that is different from nspawn itself, i.e. is using the same idmapping as nspawn, then the user namespace attached to nspawn should be reused for custom mounts as well. Not just does this safe memory it also allows systemd to skip a bunch of additional setup work.

So nspawn should probably start stashing the user namespace it uses for the idmapping somewhere. Either by bind-mounting it somewhere or by stashing the fd internally.

Reply:
I'm not trying to put this on your plate. :) But the current design is just very wasteful. It is a perfectly valid way of doing things but it wastes a lot of resources and generates unneeded work.

Reply:
the main reason this works this way is that (unlike lxc) we mount the file systems before doing the usern dance, i.e. we want the mounts to be owned by the host userns so that they are immutable to the container, following the idea that mounts established by the container runtime are owned by the host, and those established by the container payload are owned by the payload.

Of course, we could allocate the userns early on, then leave it again, useit everywhere and then later join it again, but I didn't see the big benefit of that, given this requires a non-trivial amount of complexity.